src/tiled/templatesdock.cpp

```diff
@@ -78,20 +125,46 @@ TemplatesDock::TemplatesDock(QWidget *parent):
 
     mTemplatesView->setModel(model);
 
+    mMapScene->setSelectedTool(mToolManager->selectedTool());
+
+    connect(mTemplatesView, &TemplatesView::currentTemplateChanged,
+            this, &TemplatesDock::setTemplate);
+
     connect(mTemplatesView, &TemplatesView::currentTemplateChanged,
             this, &TemplatesDock::currentTemplateChanged);
 
     connect(mTemplatesView->model(), &ObjectTemplateModel::dataChanged,
             mTemplatesView, &TemplatesView::applyTemplateGroups);
+
     connect(mTemplatesView->model(), &ObjectTemplateModel::rowsInserted,
             mTemplatesView, &TemplatesView::applyTemplateGroups);
+
+    connect(mToolManager, &ToolManager::selectedToolChanged,
+            this, &TemplatesDock::setSelectedTool);
+
+    setFocusPolicy(Qt::ClickFocus);
+    mTemplatesView->setFocusProxy(this);
+    mMapView->setFocusProxy(this);
 }
 
 TemplatesDock::~TemplatesDock()
 {
+    mMapScene->disableSelectedTool();
+
+    disconnect(mDummyMapDocument->undoStack(), &QUndoStack::indexChanged,
```

Comment:
You should only try this when `mDummyMapDocument` isn't null. On the other hand, this should also not be needed at all, because the deletion of the dummy document should delete the undo stack, which should disconnect this signal.